tests/nn/data_parallel/test_fsdp_no_sync.py

```diff
@@ -94,37 +96,82 @@ def _test_no_sync(self, model, batch_dim):
         assert objects_are_equal(ref_grads, accumulated_grads, raise_exception=True)
 
 
+keys = ["reshard_after_forward", "mixed_precision"]
+COMM_CONFIG_OPTIONS = [[dict(zip(keys, config))] for config in itertools.product([True, False], repeat=len(keys))]
+
+
 class TestNoSyncCommunication(DistributedTest):
-    def test_communication(self):
-        config = {"mixed_precision": True}
+    @parameterized.expand(COMM_CONFIG_OPTIONS, name_func=rename_test)
+    def test_communication(self, config):
         fn = functools.partial(self._test_communication, config=config)
         spawn_and_init(fn)
 
+    @parameterized.expand(COMM_CONFIG_OPTIONS, name_func=rename_test)
+    def test_communication_nested(self, config):
+        fn = functools.partial(self._test_communication, config=config, nested_model=True)
+        spawn_and_init(fn)
+
     @classmethod
-    def _test_communication(self, rank, group, config):
+    def _test_communication(self, rank, group, config, nested_model=False):
         if group.size() == 1:
             return
 
-        model = self.get_wrapped_model(group, config=config)
+        # Turn off bucketing to accurately count number of reduce_scatters.
+        config["bucket_cap_mb"] = 0
+
+        if nested_model:
+            model = NestedWrappedModule(group, config)
+            model = FullyShardedDataParallel(model, group, **config).cuda()
+        else:
+            model = self.get_wrapped_model(group, config=config)
+
+        num_fsdp = 0
+        for child in model.modules():  # includes self
+            if isinstance(child, FullyShardedDataParallel) and len(child.params) > 0:
+                num_fsdp += 1
+
+        if config.get("reshard_after_forward", True):
+            # inside no_sync:
+            #   num_fsdp all-gathers in the forward
```

Comment:
it's a bit difficult to wrap my head around, why inside no_sync the forward all gather is one more than outside no_sync's forward?